language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java

```diff
@@ -0,0 +1,105 @@
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.QualifiedNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.RequiredParameterNode;
+import io.ballerina.compiler.syntax.tree.Token;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.common.utils.completion.QNameReferenceUtil;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+import org.ballerinalang.langserver.commons.completion.LSCompletionException;
+import org.ballerinalang.langserver.commons.completion.LSCompletionItem;
+import org.ballerinalang.langserver.completions.SnippetCompletionItem;
+import org.ballerinalang.langserver.completions.providers.AbstractCompletionProvider;
+import org.ballerinalang.langserver.completions.util.Snippet;
+import org.ballerinalang.langserver.completions.util.SortingUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Handles the completions for the {@link io.ballerina.compiler.syntax.tree.RequiredParameterNode}.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.completion.spi.BallerinaCompletionProvider")
+public class RequiredParameterNodeContext extends AbstractCompletionProvider<RequiredParameterNode> {
+
+    public RequiredParameterNodeContext() {
+        super(RequiredParameterNode.class);
+    }
+
+    @Override
+    public List<LSCompletionItem> getCompletions(BallerinaCompletionContext context, RequiredParameterNode node)
+            throws LSCompletionException {
+
+        List<LSCompletionItem> completionItems = new ArrayList<>();
+        NonTerminalNode nodeAtCursor = context.getNodeAtCursor();
+        if (QNameReferenceUtil.onQualifiedNameIdentifier(context, nodeAtCursor)) {
+            /*
+                Covers the Following
+                (1) function(mod:<cursor>)
+                (2) function(mod:T<cursor>)
+            */
+            QualifiedNameReferenceNode qNameRef = (QualifiedNameReferenceNode) nodeAtCursor;
+            completionItems.addAll(this.getCompletionItemList(QNameReferenceUtil
+                    .getTypesInModule(context, qNameRef), context));
+        } else {
+            /*
+              covers the following
+              function (int,int,..) re<cursor>
+            */
+            if (node.typeName() instanceof FunctionTypeDescriptorNode && withinParamName(node, context)) {
+                Optional<FunctionSignatureNode> functionSignature =
+                        ((FunctionTypeDescriptorNode) node.typeName()).functionSignature();
+                if (functionSignature.isPresent() && functionSignature.get().returnTypeDesc().isEmpty()) {
+                    completionItems.add(new SnippetCompletionItem(context, Snippet.KW_RETURNS.get()));
+                }
+            } else {
+                /*
+                Covers the Following
+                (1) function(T<cursor>)
+                */
```

Comment:
With this approach, aren't we getting a hit to the else clause for the following case as well? 
`function (int a<cursor>)`

Reply:
Noted. We might get. Let me add a test case for that as well

Reply:
My concern is, we should not provide completions in that particular case. Currently we provide completions for that particular context right? If so shall we handle that as well?

Reply:
Have handled it.